File: client/src/test/scala/stasis/test/specs/unit/client/collection/rules/SourceUriSpec.scala
package stasis.test.specs.unit.client.collection.rules

import stasis.client.collection.rules.SourceUri
import stasis.test.specs.unit.UnitSpec

class SourceUriSpec extends UnitSpec {
  "A SourceUri" should "extract the scheme from a source" in {
    val schemeTable: Map[String, Option[String]] = Map(
      // no scheme -> local filesystem
      "/home/test" -> None,
      "home/test" -> None,
      "/home/test path" -> None,
      "" -> None,

      // not schemes: single-char prefixes, numeric prefixes
      "C:\\test" -> None,
      "12:30" -> None,

      // library schemes
      "photos:/test/file.heic" -> Some("photos"),
      "photos://test/file.heic" -> Some("photos"),
      "photos:test" -> Some("photos"),
      "photos:" -> Some("photos"),
      "contacts:/" -> Some("contacts"),

      // scheme charset ([A-Za-z][A-Za-z0-9+.-]+) and case preservation
      "x-y.z+1:/a" -> Some("x-y.z+1"),
      "PHOTOS:/x" -> Some("PHOTOS")
    )

    schemeTable.foreach { case (source, expected) =>
      withClue(s"Extracting scheme from source [$source]") {
        SourceUri.scheme(source) should be(expected)
      }
    }
  }
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/collection/rules/SourceUriSpec.kt
package stasis.test.client_android.lib.collection.rules

import io.kotest.assertions.withClue
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import stasis.client_android.lib.collection.rules.SourceUri

class SourceUriSpec : WordSpec({
    "A SourceUri" should {
        "extract the scheme from a source" {
            val schemeTable: Map<String, String?> = mapOf(
                "/home/test" to null,
                "home/test" to null,
                "/home/test path" to null,
                "" to null,

                "C:\\test" to null,
                "12:30" to null,

                "photos:/test/file.heic" to "photos",
                "photos://test/file.heic" to "photos",
                "photos:test" to "photos",
                "photos:" to "photos",
                "contacts:/" to "contacts",

                "x-y.z+1:/a" to "x-y.z+1",
                "PHOTOS:/x" to "PHOTOS"
            )

            schemeTable.forEach { (source, expected) ->
                withClue("Extracting scheme from source [$source]") {
                    SourceUri.scheme(source) shouldBe (expected)
                }
            }
        }
    }
})
